blobs.go

```diff
@@ -0,0 +1,71 @@
+package distribution
+
+import (
+	"io"
+	"net/http"
+	"time"
+
+	"github.com/docker/distribution/digest"
+	"golang.org/x/net/context"
+)
+
+// Blob describes a readable and seekable binary object. Most registry
+// objects, including layers and manifests can be accessible by digest in a
+// BlobService.
+type Blob interface {
+	Describable
+	io.ReaderAt
+	io.Closer
+
+	Reader() (ReadSeekCloser, error)
+
+	// Handler returns an HTTP handler which serves the blob content, either
+	// by redirection or directly serving the content itself.
+	Handler(r *http.Request) (http.Handler, error)
+}
+
+// BlobWriter provides a handle for working with a blob store. Instances
+// should be obtained from BlobService.Writer and BlobService.Resume. If
+// supported by the store, a writer can be recovered with the id.
+type BlobWriter interface {
+	io.WriteSeeker
+	io.WriterAt
+	io.ReaderFrom // TODO(stevvooe): This should be optional.
```

Comment:
> // TODO(stevvooe): This should be optional.

`io.ReaderFrom` is probably the most important optimization from the client perspective. Without it either making a request per write is necessary or having an open request in which Write sends to the request body.